llama-cli/src/main.rs

```diff
@@ -314,3 +317,81 @@ fn main() {
         }
     }
 }
+
+fn ad_hoc_test() {
+    let (mut model, vocab) = llama_rs::Model::load(&CLI_ARGS.model_path, 2048, |_| {}).unwrap();
+    let mut session = model.start_session(InferenceSessionParameters {
+        last_n_size: 64,
+        memory_k_type: ModelKVMemoryType::Float32,
+        memory_v_type: ModelKVMemoryType::Float32,
+    });
+
+    let inference_params = InferenceParameters {
+        n_threads: 8,
+        n_batch: 8,
+        top_k: 40,
+        top_p: 0.95,
+        repeat_penalty: 1.30,
+        temp: 0.80,
+        bias_tokens: TokenBias::default(),
+    };
+
+    session.feed_prompt::<Infallible>(
+        &model,
+        &vocab,
+        &inference_params,
+        "My favourite animal is the ",
+        |_| Ok(()),
+    );
+
+    let mut output_request = EvaluateOutputRequest {
+        all_logits: None,
+        embeddings: Some(Vec::new()),
+    };
+
+    let dog = model.tokenize(&vocab, "dog", false).unwrap();
+    model.evaluate(&mut session, 8, &dog, &mut output_request);
+
+    println!("Embeddings for 'dog' (1):");
+    for x in output_request.embeddings.as_ref().unwrap() {
+        //print!("{x:.2} ")
+    }
+    println!();
+
+    let mut session2 = model.start_session(InferenceSessionParameters {
+        last_n_size: 64,
+        memory_k_type: ModelKVMemoryType::Float32,
+        memory_v_type: ModelKVMemoryType::Float32,
+    });
+    session2.feed_prompt::<Infallible>(
+        &model,
+        &vocab,
+        &inference_params,
+        "I have just adopted a cute ",
+        |_| Ok(()),
+    );
+    let mut output_request2 = EvaluateOutputRequest {
+        all_logits: None,
+        embeddings: Some(Vec::new()),
+    };
+
+    // Try other words: 'dog', 'cat', 'potato', '$' -> To see decreasingly lower dot product values.
+    let dog2 = model.tokenize(&vocab, "dog", false).unwrap();
```

Comment:
What I'm doing here is feeding the following two sentences through the transformer:

- "My favourite animal is the dog"
- "I just adopted a cute dog"

Afterwards, I retrieve the embeddings for the last token (dog), and compute their similarity with a simple dot product.

Then, I tried changing the second sentence from 'dog' to 'cat', 'potato', '$' respectively, and the semantic similarity dropped accordingly, with $ ranking the lowest.

Reply:
@setzer22 will feed prmpt before eval has different embeddings compared to eval all tokens together?

Reply:
@hlhr202 The embeddings wouldn't be affected, but you shouldn't call `evaluate` with the whole prompt like that for a couple reasons:

- A call to evaluate runs all the tokens you give it as a batch, meaning it requires increased memory usage. For very long prompts, this could become very expensive.
- The output value will return the output embeddings for every token that you fed through `eval`. This means you would be retrieving a lot more embedding data than for just the word "dog".

Reply:
@setzer22 I just understand your comments here. this means we can only extract embeddings for a single part of words (which may also have hidden information mixed with context of the whole sentence). that should a little bit different with OpenAI's embedding function. what i understand for openai's embedding, is for the whole sentence but at the same time returned in a fixed size of tensor... that is quite beyond my knowledge though

Reply:
well I guess I might find possible ways to implement such 'sentence embedding', I will try add some special end token and extract the hidden layer once the end token evaluated. not sure if it works, but it must worth a try.